near/src/runtime.rs

```diff
@@ -599,51 +599,31 @@ impl RuntimeAdapter for NightshadeRuntime {
         block_index: BlockIndex,
         block_timestamp: u64,
         gas_price: Balance,
-        state_root: StateRoot,
-        transaction: SignedTransaction,
-    ) -> Result<ValidTransaction, RuntimeError> {
-        let mut state_update = TrieUpdate::new(self.trie.clone(), state_root.hash);
+        state_update: &mut TrieUpdate,
+        transaction: &SignedTransaction,
+    ) -> Result<Gas, RuntimeError> {
         let apply_state = ApplyState {
             block_index,
             epoch_length: self.genesis_config.epoch_length,
             gas_price,
             block_timestamp,
         };
 
-        if let Err(err) = self.runtime.verify_and_charge_transaction(
-            &mut state_update,
-            &apply_state,
-            &transaction,
-        ) {
-            debug!(target: "runtime", "Tx {:?} validation failed: {:?}", transaction, err);
-            return Err(err);
-        }
-        Ok(ValidTransaction { transaction })
+        self.runtime
+            .verify_and_charge_transaction(state_update, &apply_state, &transaction)
+            .map_err(|err| {
+                state_update.rollback();
+                debug!(target: "runtime", "Tx {:?} validation failed: {:?}", transaction, err);
+                err
+            })
+            .map(|verification_result| {
+                state_update.commit();
+                verification_result.gas_burnt
+            })
     }
 
-    fn filter_transactions(
-        &self,
-        block_index: BlockIndex,
-        block_timestamp: u64,
-        gas_price: Balance,
-        state_root: StateRoot,
-        transactions: Vec<SignedTransaction>,
-    ) -> Vec<SignedTransaction> {
-        let mut state_update = TrieUpdate::new(self.trie.clone(), state_root.hash);
-        let apply_state = ApplyState {
-            block_index,
-            epoch_length: self.genesis_config.epoch_length,
-            gas_price,
-            block_timestamp,
-        };
-        transactions
-            .into_iter()
-            .filter(|transaction| {
-                self.runtime
-                    .verify_and_charge_transaction(&mut state_update, &apply_state, transaction)
-                    .is_ok()
-            })
-            .collect()
+    fn get_state_update(&self, state_root: StateRoot) -> TrieUpdate {
```

Comment:
really don't like this idea of exposing TrieUpdate outside of the runtime

Reply:
Suggestions?

Reply:
@nearmax suggested pass draining iterator to the the runtime. It will also have to take a filter function from the client for verifying `blockhash` from the tx to be on the same chain. I'll try to rewrite it.